File: rulebook-cppcheck/rulebook_cppcheck/checkers/lowercase_hex.py
from typing import override

from rulebook_cppcheck.checkers.rulebook_checkers import RulebookTokenChecker
from rulebook_cppcheck.messages import _Messages

try:
    from cppcheckdata import Token
except ImportError:
    from cppcheck.Cppcheck.addons.cppcheckdata import Token


class LowercaseHexChecker(RulebookTokenChecker):
    """See detail: https://hanggrian.github.io/rulebook/rules/#lowercase-hex"""
    ID: str = 'lowercase-hex'
    _MSG: str = 'lowercase.hex'

    @override
    def process_tokens(self, tokens: list[Token]) -> None:
        # checks for violation
        for token in [t for t in tokens if t.isNumber]:
            value: str = token.str
            if not value.lower().startswith('0x'):
                return
            value_replacement: str = value.lower()
            if value == value_replacement:
                return
            self.report_error(token, _Messages.get(self._MSG, value_replacement))

File: rulebook-eslint/src/rules/lowercase-hex.ts
import { Rule } from 'eslint';
import { Literal } from 'estree';
import messages from '../messages.js';
import RulebookRule from './rulebook-rules.js';

/** {@link https://hanggrian.github.io/rulebook/rules/#lowercase-hex|See detail} */
class LowercaseHexRule extends RulebookRule {
    constructor() {
        super('lowercase-hex', {
            [LowercaseHexRule.MSG]: messages.get(LowercaseHexRule.MSG),
        });
    }

    create(context: Rule.RuleContext) {
        return {
            Literal(node: Literal) {
                // checks for violation
                if (typeof node.value !== 'number') {
                    return;
                }
                const value = node.raw;
                if (value === undefined || !value.toLowerCase().startsWith('0x')) {
                    return;
                }
                const valueReplacement = value.toLowerCase();
                if (value === valueReplacement) {
                    return;
                }
                context.report({
                    node,
                    messageId: LowercaseHexRule.MSG,
                    data: {
                        $0: valueReplacement,
                    },
                });
            },
        };
    }

    static MSG: string = 'lowercase.hex';
}

export default new LowercaseHexRule();
